src/clojure_lsp/crawler.clj

```diff
@@ -67,6 +67,32 @@
        :message (str "Unknown forward declaration: " (:str usage))
        :severity 1})))
 
+(defn ^:private supports-argc [signature argc]
+  (let [min-argc (count (take-while #(not= '& %) signature))
+        has-rest (not= min-argc (count signature))]
+    (if has-rest
+      (>= argc min-argc)
+      (= argc min-argc))))
+
+(defn ^:private diagnose-wrong-arity [uri usages]
+  (let [all-envs (assoc (:file-envs @db/db) uri usages)
+        function-references (filter :signatures usages)
+        call-sites (->> all-envs
+                        (mapcat (comp val))
+                        (filter :argc))]
+    (for [call-site call-sites
+          :let [argc (:argc call-site)
+                function-sym (:sym call-site)
+                relevant-functions (filter #(->> % (:sym) (= function-sym)) function-references)]
+          :when (when-let [relevant-function (last relevant-functions)]
+                  (not-any? #(supports-argc % argc) (:signatures relevant-function)))]
+      {:range (shared/->range call-site)
+       :code :wrong-arity
+       :message (let [plural (not= argc 1)]
+                  (format "No overload %s for %d argument%s"
+                          (:str call-site) argc (if plural "s" "")))
+       :severity 1})))
+
```

Comment:
I loaded up your branch and I was curious why clojure.core fields weren't being checked for arity.

It's because `usages` and `all-envs` are mixed up, you want the `call-sites` from `usages` (the current doc) and `function-references` from `all-envs`.  Then I found that many of the `clojure.core` signatures are returning as `(quote ([a] [a b]))`, or have type hints like `'(^clojure.lang.ISeq [coll])` which explode in `supports-argc`. So I added a `catch` to ignore functions it can't handle. 

Here's where I am but there's still some functions erroneously being flagged. For instance, in this project `zsub/subzip` takes 1 arg but it's diagnosed as wrong arity.



```
(defn ^:private diagnose-wrong-arity [uri usages]
  (let [all-envs (assoc (:file-envs @db/db) uri usages)
        function-references (filter :signatures (mapcat val all-envs))
        call-sites (->> usages
                        (filter :argc))]
    (for [call-site call-sites
          :let [argc (:argc call-site)
                function-sym (:sym call-site)
                relevant-functions (filter #(->> % (:sym) (= function-sym)) function-references)]
          :when (when-let [relevant-function (last relevant-functions)]
                  (:sym relevant-function)
                  (try
                    (not-any? #(supports-argc % argc) (:signatures relevant-function))
                    (catch Exception e
                      false)))]
      {:range (shared/->range call-site)
       :code :wrong-arity
       :message (let [plural (not= argc 1)]
                  (format "No overload %s for %d argument%s"
                          (:str call-site) argc (if plural "s" "")))
       :severity 1})))
```

Reply:
Thanks for the review!

I've applied this usages/all-envs fix, and also I reworked the :signatures storage to remember both the arglists and the eval'd signature, although this broke one of the parser tests somehow.

I will try and look at fixing the broken test tomorrow.

Reply:
Okay most of the issues with annotations and quoted arglists are fixed now, however some macros like `defn` have `:arglists` written almost like a regex, which `supports-argc` just interprets literally for now, which is broken.

I'm not sure if the best approach is to have an ignore list for functions like these or to auto-detect if an `:arglists` looks like a regex or something else

Reply:
An ignore list would be fine - I don't think clojure.core itself is super consistent about those annotations. 

Another possibility would be a map that allowed overriding the signatures or argcs for a function.

Reply:
> I wasn’t able to reproduce the `subzip` misdiagnosis

It's possible that was a repl state issue, it was in parser.clj. But I'll give it another look once these changes are ready.

Reply:
Okay I've added a short ignore list. Checking over parser.clj I saw an issue with nested threading but that should be fixed now, and I've merged from master so things should be ready now.